File: backend/src/test/java/lan/dk/podcastserver/manager/worker/tf1replay/TF1ReplayFinderTest.java
package lan.dk.podcastserver.manager.worker.tf1replay;

import com.github.davinkevin.podcastserver.service.HtmlService;
import com.github.davinkevin.podcastserver.service.ImageService;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import com.github.davinkevin.podcastserver.IOUtils;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.io.IOException;
import java.net.URISyntaxException;

import static lan.dk.podcastserver.assertion.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.when;

/**
 * Created by kevin on 21/07/2016.
 */
@RunWith(MockitoJUnitRunner.class)
public class TF1ReplayFinderTest {

    private @Mock HtmlService htmlService;
    private @Mock ImageService imageService;
    private @InjectMocks
    TF1ReplayFinder finder;

    @Test
    public void should_fetch_from_html_page() throws IOException, URISyntaxException {
        /* Given */
        String url = "www.tf1.fr/tf1/19h-live/videos";
        when(htmlService.get(eq(url))).thenReturn(IOUtils.fileAsHtml("/remote/podcast/tf1replay/19h-live.html"));
        when(imageService.getCoverFromURL(anyString())).then(i -> Cover.builder().url(i.getArgument(0)).build());

        /* When */
        Podcast podcast = finder.find(url);

        /* Then */
        assertThat(podcast)
                .hasTitle("Vidéos & Replay 19h live - TF1")
                .hasDescription("Tous les replays  19h live: les vidéos bonus exclusives des coulisses, des interviews de  19h live:")
                .hasUrl(url)
                .hasType("TF1Replay");

        assertThat(podcast.getCover()).hasUrl("http://photos1.tf1.fr/1920/960/1920x1080-19h-5619b8-0@1x.jpg");
    }

    @Test
    public void should_fetch_from_html_page_without_url() throws IOException, URISyntaxException {
        /* Given */
        String url = "www.tf1.fr/tf1/19h-live/videos";
        when(htmlService.get(eq(url))).thenReturn(IOUtils.fileAsHtml("/remote/podcast/tf1replay/19h-live.withoutpicture.html"));
        when(imageService.getCoverFromURL(anyString())).then(i -> Cover.builder().url(i.getArgument(0)).build());

        /* When */
        Podcast podcast = finder.find(url);

        /* Then */
        assertThat(podcast)
                .hasTitle("Vidéos & Replay 19h live - TF1")
                .hasDescription("Tous les replays  19h live: les vidéos bonus exclusives des coulisses, des interviews de  19h live:")
                .hasUrl(url)
                .hasType("TF1Replay");

        assertThat(podcast.getCover()).hasUrl("http://photos2.tf1.fr/130/65/logo_programme-284-3955bf-0@1x.jpg");
    }

    @Test
    public void should_be_compatible() {
        /* Given */
        String url = "www.tf1.fr/tf1/19h-live/videos";
        /* When */
        Integer compatibility = finder.compatibility(url);
        /* Then */
        assertThat(compatibility).isEqualTo(1);
    }

    @Test
    public void should_not_be_compatible() {
        /* Given */
        String url = "www.tf1.com/foo/bar/videos";
        /* When */
        Integer compatibility = finder.compatibility(url);
        /* Then */
        assertThat(compatibility).isEqualTo(Integer.MAX_VALUE);
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/manager/worker/tf1replay/TF1ReplayFinderTest.kt
package com.github.davinkevin.podcastserver.manager.worker.tf1replay

import arrow.core.None
import com.github.davinkevin.podcastserver.IOUtils.fileAsHtml
import com.github.davinkevin.podcastserver.entity.Cover
import com.github.davinkevin.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.utils.toVΛVΓ
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Import
import org.springframework.test.context.junit.jupiter.SpringExtension

/**
 * Created by kevin on 21/07/2016.
 */
@ExtendWith(SpringExtension::class)
class TF1ReplayFinderTest(
    @Autowired val htmlService: HtmlService,
    @Autowired val imageService: ImageService,
    @Autowired val finder: TF1ReplayFinder
) {

    @Test
    fun `should fetch from html page`() {
        /* Given */
        val anUrl = "https://www.tf1.fr/tmc/quotidien-avec-yann-barthes/videos/replay"
        whenever(htmlService.get(anUrl)).thenReturn(fileAsHtml("/remote/podcast/tf1replay/quotidien.root.html"))
        whenever(imageService.getCoverFromURL(any())).then { i -> Cover().apply { url = i.getArgument(0) } }

        /* When */
        val podcast = finder.find(anUrl)

        /* Then */
        assertThat(podcast.title).isEqualTo("Quotidien avec Yann Barthès - TMC | MYTF1")
        assertThat(podcast.description).isEqualTo("Yann Barthès est désormais sur TMC et TF1")
        assertThat(podcast.url).isEqualTo(anUrl)
        assertThat(podcast.type).isEqualTo("TF1Replay")
        assertThat(podcast.cover!!.url).isEqualTo("https://photos.tf1.fr/1200/0/vignette-portrait-quotidien-2-aa530a-0@1x.png")
    }

    @Test
    fun `should fetch from html page without url`() {
        /* Given */
        val anUrl = "https://www.tf1.fr/tmc/quotidien-avec-yann-barthes/videos/replay"
        whenever(htmlService.get(anUrl)).thenReturn(fileAsHtml("/remote/podcast/tf1replay/quotidien.root-without-picture.html"))
        whenever(imageService.getCoverFromURL(any())).then { i -> Cover().apply { url = i.getArgument(0) } }

        /* When */
        val podcast = finder.find(anUrl)

        /* Then */
        assertThat(podcast.title).isEqualTo("Quotidien avec Yann Barthès - TMC | MYTF1")
        assertThat(podcast.description).isEqualTo("Yann Barthès est désormais sur TMC et TF1")
        assertThat(podcast.url).isEqualTo(anUrl)
        assertThat(podcast.type).isEqualTo("TF1Replay")
        assertThat(podcast.cover!!.url).isNull()
    }

    @Test
    fun `should return default podcast if request end up without result`() {
        /* Given */
        val anUrl = "https://www.tf1.fr/tmc/quotidien-avec-yann-barthes/videos/replay"
        whenever(htmlService.get(anUrl)).thenReturn(None.toVΛVΓ())

        /* When */
        val podcast = finder.find(anUrl)

        /* Then */
        assertThat(podcast).isSameAs(Podcast.DEFAULT_PODCAST)
    }

    @Test
    fun `should have default cover if nothing comming back from backend server`() {
        /* Given */
        val anUrl = "https://www.tf1.fr/tmc/quotidien-avec-yann-barthes/videos/replay"
        whenever(htmlService.get(anUrl)).thenReturn(fileAsHtml("/remote/podcast/tf1replay/quotidien.root.html"))
        whenever(imageService.getCoverFromURL(any())).thenReturn(null)

        /* When */
        val podcast = finder.find(anUrl)

        /* Then */
        assertThat(podcast.title).isEqualTo("Quotidien avec Yann Barthès - TMC | MYTF1")
        assertThat(podcast.description).isEqualTo("Yann Barthès est désormais sur TMC et TF1")
        assertThat(podcast.url).isEqualTo(anUrl)
        assertThat(podcast.type).isEqualTo("TF1Replay")
        assertThat(podcast.cover).isSameAs(Cover.DEFAULT_COVER)

    }

    @Test
    fun `should be compatible`() {
        /* Given */
        val url = "https://www.tf1.fr/tmc/quotidien-avec-yann-barthes/videos/replay"
        /* When */
        val compatibility = finder.compatibility(url)
        /* Then */
        assertThat(compatibility).isEqualTo(1)
    }

    @Test
    fun `should not be compatible`() {
        /* Given */
        val url = "www.tf1.com/foo/bar/videos"
        /* When */
        val compatibility = finder.compatibility(url)
        /* Then */
        assertThat(compatibility).isEqualTo(Integer.MAX_VALUE)
    }

    @TestConfiguration
    @Import(TF1ReplayFinder::class)
    class LocalTestConfiguration {
        @Bean fun htmlService() = mock<HtmlService>()
        @Bean fun imageService() = mock<ImageService>()
    }
}
